File: src/omnicontext/commands/uninstall.py
import os

from omnicontext.hooks import get_git_root, uninstall_hook


def cmd_uninstall(args):
    if "--global" in args:
        os.system("git config --global --unset core.hooksPath")
        print("Global hooks path unset")
        return 0

    git_root = get_git_root()
    if not git_root:
        print("error: not a git repository")
        return 1

    result = uninstall_hook(git_root)

    if result == "uninstalled":
        print("Hook removed")
        return 0
    elif result == "not_installed":
        print("No hook installed")
        return 0
    elif result == "not_managed":
        print("error: hook exists but not managed by omnicontext")
        return 1

    return 1

File: apps/cli/src/commands/uninstall.ts
import {
  CLI_NAME,
  HOOK_POST_CHECKOUT,
  HOOK_POST_COMMIT,
  HookType,
  HookUninstallResult,
  uninstallHook,
} from '@branch-context/core';
import type { Program } from '@caporal/core';
import { requireGitRoot } from '../helpers/git-root';

const hookUninstallMessages = {
  [HookUninstallResult.Uninstalled]: (hookName: string) => `Hook removed: ${hookName}`,
  [HookUninstallResult.NotManaged]: (hookName: string) =>
    `warning: ${hookName} hook exists but not managed by ${CLI_NAME}`,
  [HookUninstallResult.NotInstalled]: () => null,
} as const satisfies Record<HookUninstallResult, (hookName: string) => string | null>;

export function registerUninstallCommand(program: Program) {
  program.command('uninstall', 'Remove hooks from current repo').action(() => cmdUninstall());
}

function cmdUninstall() {
  const gitRoot = requireGitRoot();
  if (!gitRoot) {
    return 1;
  }

  const checkoutResult = uninstallHook(gitRoot, HookType.PostCheckout);
  const commitResult = uninstallHook(gitRoot, HookType.PostCommit);

  printHookUninstallResult(checkoutResult, HOOK_POST_CHECKOUT);
  printHookUninstallResult(commitResult, HOOK_POST_COMMIT);

  if (
    checkoutResult === HookUninstallResult.NotInstalled &&
    commitResult === HookUninstallResult.NotInstalled
  ) {
    console.log('No hooks installed');
  }

  return 0;
}

function printHookUninstallResult(result: HookUninstallResult, hookName: string) {
  const message = hookUninstallMessages[result](hookName);
  if (message) {
    console.log(message);
  }
}
